packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,707 @@ import 'picker.dart';
 /// Default aesthetic values obtained by comparing with iOS pickers.
 const double _kItemExtent = 32.0;
 const double _kPickerWidth = 330.0;
+const bool _kUseMagnifier = true;
+const double _kMagnification = 1.1;
+const double _kDatePickerPadSize = 12.0;
 /// Considers setting the default background color from the theme, in the future.
 const Color _kBackgroundColor = CupertinoColors.white;
 
 
+// Lays out the date picker based on how much space each single column needs.
+//
+// Each column will be padded horizontally by 12.0 both left and right.
+//
+// The picker will be placed in the center, and the leftmost and rightmost
+// column will be extended equally to the remaining width.
+class _DatePickerLayout extends MultiChildLayoutDelegate {
+  _DatePickerLayout({
+    @required this.columnWidth,
+    @required this.textDirectionFactor,
+  }) : assert(columnWidth != null),
+       assert(textDirectionFactor != null);
+
+  // The list containing widths of all columns.
+  final List<double> columnWidth;
+
+  // textDirectionFactor is 1 if text is written left to right, and -1 if right to left.
+  final int textDirectionFactor;
+
+  @override
+  void performLayout(Size size) {
+    double remainingWidth = size.width;
+
+    for (int i = 0; i < columnWidth.length; i++)
+      remainingWidth -= columnWidth[i] + _kDatePickerPadSize * 2;
+
+    double currentHorizontalOffset = 0.0;
+
+    for (int i = 0; i < columnWidth.length; i++) {
+      final int index = textDirectionFactor == 1 ? i : columnWidth.length - i - 1;
+
+      double childWidth = columnWidth[index] + _kDatePickerPadSize * 2;
+      if (index == 0 || index == columnWidth.length - 1)
+        childWidth += remainingWidth / 2;
+
+      layoutChild(index, BoxConstraints.tight(new Size(childWidth, size.height)));
+      positionChild(index, new Offset(currentHorizontalOffset, 0.0));
+
+      currentHorizontalOffset += childWidth;
+    }
+  }
+
+  @override
+  bool shouldRelayout(_DatePickerLayout oldDelegate) {
+    return columnWidth != oldDelegate.columnWidth
+      || textDirectionFactor != oldDelegate.textDirectionFactor;
+  }
+}
+
+/// Different modes of [CupertinoDatePicker].
```

Comment:
Say that the mode shouldn't change over the lifespan of this widget. Assert in didUpdateWidget.

Reply:
If we're imposing constraints on how to use the API in the documentation, it may be a good idea to rethink the API surface we expose, since it may last a very long time.

This is just a thought, but it might be easier to use if we made the enum private and used named constructors for the different behavior modes.